Purpose:
Predict if a person will default their credit card next month

In [1]:
import pandas as pd
df=pd.read_csv('D:/Dropbox/Machine Learning/Data/Credit Default/UCI_Credit_Card.csv')
# df=pd.read_csv('/Users/jiahuali1991/Dropbox/Machine Learning/Data/Credit Default/UCI_Credit_Card.csv')
df.sample(5)

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  \
4578    4579    40000.0    2          2         2   23     -1      2     -1   
25877  25878   200000.0    2          1         2   47     -1     -1     -1   
18161  18162    20000.0    1          2         2   32      0      0      0   
7961    7962    20000.0    1          2         2   25      4      5      4   
7205    7206   130000.0    2          2         2   25      0      0      0   

       PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
4578      -1  ...    26370.0    24572.0    24918.0       0.0     326.0   
25877     -1  ...   133073.0    45373.0      973.0    2061.0  175000.0   
18161     -1  ...     8115.0     2594.0     1610.0    1306.0    3045.0   
7961       3  ...    18248.0    17816.0    18021.0       0.0       0.0   
7205       0  ...   130239.0   127980.0   130305.0    4023.0    4030.0   

       PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default.payment.next.month  
4578    26370.0     886.0    1075.0     830.0                           1  
25877  133100.0   46000.0    1000.0       0.0                           0  
18161   10000.0       0.0    1610.0       0.0                           0  
7961        0.0       0.0     636.0    1200.0                           1  
7205    20000.0       0.0    4500.0       0.0                           1  

[5 rows x 25 columns]

Dataset Information This dataset contains information on default payments, demographic factors, credit data, history of payment, and bill statements of credit card clients in Taiwan from April 2005 to September 2005.

Content There are 25 variables:

ID: ID of each client LIMIT_BAL: Amount of given credit in NT dollars (includes individual and family/supplementary credit SEX: Gender (1=male, 2=female) EDUCATION: (1=graduate school, 2=university, 3=high school, 4=others, 5=unknown, 6=unknown) MARRIAGE: Marital status (1=married, 2=single, 3=others) AGE: Age in years PAY_0: Repayment status in September, 2005 (-1=pay duly, 1=payment delay for one month, 2=payment delay for two months, … 8=payment delay for eight months, 9=payment delay for nine months and above) PAY_2: Repayment status in August, 2005 (scale same as above) PAY_3: Repayment status in July, 2005 (scale same as above) PAY_4: Repayment status in June, 2005 (scale same as above) PAY_5: Repayment status in May, 2005 (scale same as above) PAY_6: Repayment status in April, 2005 (scale same as above) BILL_AMT1: Amount of bill statement in September, 2005 (NT dollar) BILL_AMT2: Amount of bill statement in August, 2005 (NT dollar) BILL_AMT3: Amount of bill statement in July, 2005 (NT dollar) BILL_AMT4: Amount of bill statement in June, 2005 (NT dollar) BILL_AMT5: Amount of bill statement in May, 2005 (NT dollar) BILL_AMT6: Amount of bill statement in April, 2005 (NT dollar) PAY_AMT1: Amount of previous payment in September, 2005 (NT dollar) PAY_AMT2: Amount of previous payment in August, 2005 (NT dollar) PAY_AMT3: Amount of previous payment in July, 2005 (NT dollar) PAY_AMT4: Amount of previous payment in June, 2005 (NT dollar) PAY_AMT5: Amount of previous payment in May, 2005 (NT dollar) PAY_AMT6: Amount of previous payment in April, 2005 (NT dollar) default.payment.next.month: Default payment (1=yes, 0=no) Inspiration Some ideas for exploration:

How does the probability of default payment vary by categories of different demographic variables? Which variables are the strongest predictors of default payment? Acknowledgements Any publications based on this dataset should acknowledge the following:

Lichman, M. (2013). UCI Machine Learning Repository [http://archive.ics.uci.edu/ml]. Irvine, CA: University of California, School of Information and Computer Science.

The original dataset can be found here at the UCI Machine Learning Repository.

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   ID                          30000 non-null  int64  
 1   LIMIT_BAL                   30000 non-null  float64
 2   SEX                         30000 non-null  int64  
 3   EDUCATION                   30000 non-null  int64  
 4   MARRIAGE                    30000 non-null  int64  
 5   AGE                         30000 non-null  int64  
 6   PAY_0                       30000 non-null  int64  
 7   PAY_2                       30000 non-null  int64  
 8   PAY_3                       30000 non-null  int64  
 9   PAY_4                       30000 non-null  int64  
 10  PAY_5                       30000 non-null  int64  
 11  PAY_6                       30000 non-null  int64  
 12  BILL_AMT1                   30000 non-null  float64
 13  BILL_AMT2                   300

In [3]:
# check missing value
df.isnull().sum()

ID                            0
LIMIT_BAL                     0
SEX                           0
EDUCATION                     0
MARRIAGE                      0
AGE                           0
PAY_0                         0
PAY_2                         0
PAY_3                         0
PAY_4                         0
PAY_5                         0
PAY_6                         0
BILL_AMT1                     0
BILL_AMT2                     0
BILL_AMT3                     0
BILL_AMT4                     0
BILL_AMT5                     0
BILL_AMT6                     0
PAY_AMT1                      0
PAY_AMT2                      0
PAY_AMT3                      0
PAY_AMT4                      0
PAY_AMT5                      0
PAY_AMT6                      0
default.payment.next.month    0
dtype: int64

In [4]:
# categorical variables description
df[['SEX','EDUCATION','MARRIAGE']].describe()

SEX     EDUCATION      MARRIAGE
count  30000.000000  30000.000000  30000.000000
mean       1.603733      1.853133      1.551867
std        0.489129      0.790349      0.521970
min        1.000000      0.000000      0.000000
25%        1.000000      1.000000      1.000000
50%        2.000000      2.000000      2.000000
75%        2.000000      2.000000      2.000000
max        2.000000      6.000000      3.000000

In [5]:
df['SEX'].value_counts(dropna=False)

2    18112
1    11888
Name: SEX, dtype: int64

In [6]:
df['EDUCATION'].value_counts(dropna=False)

2    14030
1    10585
3     4917
5      280
4      123
6       51
0       14
Name: EDUCATION, dtype: int64

SEX: Gender (1=male, 2=female) 
18112 female and 11888 male

EDUCATION: (1=graduate school, 2=university, 3=high school, 4=others, 5=unknown, 6=unknown) 
University: 14040
Graduate School: 10585
High School:4917
Unknown: 0, 5 and 6 to be combined (280+51+14)

MARRIAGE: Marital status (1=married, 2=single, 3=others)

EDUCTAION: undocumented 0 =>solution: change 0 to 5, change 6 to 5

MARRIAGE: undocumented 0 =>solution: change 0 to 3

In [7]:
df['SEX'] = df['SEX'].replace(1,'male').replace(2,'female')
df['SEX'].value_counts(dropna=False)

female    18112
male      11888
Name: SEX, dtype: int64

In [8]:
df['EDUCATION'] = df['EDUCATION'].replace(0,5).replace(6,5).replace(1,'graduate school').replace(2,'university').replace(3,'high school').replace(4,'others').replace(5,'unknown')
df['EDUCATION'].value_counts(dropna=False)


university         14030
graduate school    10585
high school         4917
unknown              345
others               123
Name: EDUCATION, dtype: int64

In [9]:
df['MARRIAGE'] = df['MARRIAGE'].replace(0,3).replace(1,'married').replace(2,'single').replace(3,'others')
df['MARRIAGE'].value_counts(dropna=False)

single     15964
married    13659
others       377
Name: MARRIAGE, dtype: int64

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   ID                          30000 non-null  int64  
 1   LIMIT_BAL                   30000 non-null  float64
 2   SEX                         30000 non-null  object 
 3   EDUCATION                   30000 non-null  object 
 4   MARRIAGE                    30000 non-null  object 
 5   AGE                         30000 non-null  int64  
 6   PAY_0                       30000 non-null  int64  
 7   PAY_2                       30000 non-null  int64  
 8   PAY_3                       30000 non-null  int64  
 9   PAY_4                       30000 non-null  int64  
 10  PAY_5                       30000 non-null  int64  
 11  PAY_6                       30000 non-null  int64  
 12  BILL_AMT1                   30000 non-null  float64
 13  BILL_AMT2                   300

In [11]:
# payment delay description
df[['PAY_0','PAY_2','PAY_3','PAY_4','PAY_5','PAY_6']].describe()

PAY_0         PAY_2         PAY_3         PAY_4         PAY_5  \
count  30000.000000  30000.000000  30000.000000  30000.000000  30000.000000   
mean      -0.016700     -0.133767     -0.166200     -0.220667     -0.266200   
std        1.123802      1.197186      1.196868      1.169139      1.133187   
min       -2.000000     -2.000000     -2.000000     -2.000000     -2.000000   
25%       -1.000000     -1.000000     -1.000000     -1.000000     -1.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max        8.000000      8.000000      8.000000      8.000000      8.000000   

              PAY_6  
count  30000.000000  
mean      -0.291100  
std        1.149988  
min       -2.000000  
25%       -1.000000  
50%        0.000000  
75%        0.000000  
max        8.000000

PAY_0: Repayment status in September, 2005 (-1=pay duly, 1=payment delay for one month, 2=payment delay for two months, … 8=payment delay for eight months, 9=payment delay for nine months and above)

PAY_?:undocumented -2

In [12]:
# bill statement description
df[['BILL_AMT1','BILL_AMT2','BILL_AMT3','BILL_AMT4','BILL_AMT5','BILL_AMT6']].describe()

BILL_AMT1      BILL_AMT2     BILL_AMT3      BILL_AMT4  \
count   30000.000000   30000.000000  3.000000e+04   30000.000000   
mean    51223.330900   49179.075167  4.701315e+04   43262.948967   
std     73635.860576   71173.768783  6.934939e+04   64332.856134   
min   -165580.000000  -69777.000000 -1.572640e+05 -170000.000000   
25%      3558.750000    2984.750000  2.666250e+03    2326.750000   
50%     22381.500000   21200.000000  2.008850e+04   19052.000000   
75%     67091.000000   64006.250000  6.016475e+04   54506.000000   
max    964511.000000  983931.000000  1.664089e+06  891586.000000   

           BILL_AMT5      BILL_AMT6  
count   30000.000000   30000.000000  
mean    40311.400967   38871.760400  
std     60797.155770   59554.107537  
min    -81334.000000 -339603.000000  
25%      1763.000000    1256.000000  
50%     18104.500000   17071.000000  
75%     50190.500000   49198.250000  
max    927171.000000  961664.000000

BILL_AMT1: Amount of bill statement in September, 2005 (NT dollar)

negative values observed: treat it as credit?

In [13]:
# previous payment description
df[['PAY_AMT1','PAY_AMT2','PAY_AMT3','PAY_AMT4','PAY_AMT5','PAY_AMT6']].describe()

PAY_AMT1      PAY_AMT2      PAY_AMT3       PAY_AMT4  \
count   30000.000000  3.000000e+04   30000.00000   30000.000000   
mean     5663.580500  5.921163e+03    5225.68150    4826.076867   
std     16563.280354  2.304087e+04   17606.96147   15666.159744   
min         0.000000  0.000000e+00       0.00000       0.000000   
25%      1000.000000  8.330000e+02     390.00000     296.000000   
50%      2100.000000  2.009000e+03    1800.00000    1500.000000   
75%      5006.000000  5.000000e+03    4505.00000    4013.250000   
max    873552.000000  1.684259e+06  896040.00000  621000.000000   

            PAY_AMT5       PAY_AMT6  
count   30000.000000   30000.000000  
mean     4799.387633    5215.502567  
std     15278.305679   17777.465775  
min         0.000000       0.000000  
25%       252.500000     117.750000  
50%      1500.000000    1500.000000  
75%      4031.500000    4000.000000  
max    426529.000000  528666.000000

In [14]:
df.LIMIT_BAL.describe()

count      30000.000000
mean      167484.322667
std       129747.661567
min        10000.000000
25%        50000.000000
50%       140000.000000
75%       240000.000000
max      1000000.000000
Name: LIMIT_BAL, dtype: float64

LIMIT_BAL: Amount of given credit in NT dollars (includes individual and family/supplementary credit


In [15]:
# rename two columns
df=df.rename(columns={'default.payment.next.month':'def_pay',
                     'PAY_0':'PAY_1'})
df.columns

Index(['ID', 'LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'PAY_1',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2',
       'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1',
       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6', 'def_pay'],
      dtype='object')

In [16]:
# Calculate default rate
df.def_pay.sum()/len(df.def_pay)

0.2212

In [17]:
# prepare X and y for machine learning
y=df['def_pay'].copy()
y.sample(5)

20006    0
20557    0
23482    1
9381     0
2089     1
Name: def_pay, dtype: int64

In [20]:
'''features=['LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'PAY_1',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2',
       'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1',
       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6']'''
#X = df[features].copy()
X = df.drop(['def_pay','ID'],axis=1)
X.columns

Index(['LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'PAY_1', 'PAY_2',
       'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2',
       'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1',
       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6'],
      dtype='object')

In [21]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler


#categorical_cols = ['SEX', 'EDUCATION', 'MARRIGE']
#numerical_cols = ['LIMIT_BAL', 'AGE', 'PAY_1',
#       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2',
#       'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1',
#       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6']

# we can achive the same goal by 
# determine categorical and numerical features
numerical_cols = X.select_dtypes(include=['int64', 'float64']).columns
categorical_cols = X.select_dtypes(include=['object', 'bool']).columns


# Preprocessing for numerical data
numerical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('standardize', StandardScaler())])

# Preprocessing for categorical data
categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])


# Bundle preprocessing for numerical and categorical data
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, numerical_cols),
        ('cat', categorical_transformer, categorical_cols)
    ])


In [22]:
# split data into training and test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.3,random_state=42)

In [23]:
df.def_pay.describe()

count    30000.000000
mean         0.221200
std          0.415062
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max          1.000000
Name: def_pay, dtype: float64

In [24]:
y_train.describe()

count    21000.000000
mean         0.222667
std          0.416046
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max          1.000000
Name: def_pay, dtype: float64

In [25]:
y_test.describe()

count    9000.000000
mean        0.217778
std         0.412758
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max         1.000000
Name: def_pay, dtype: float64

# Dummy Classifier

In [26]:
from sklearn.dummy import DummyClassifier
clf = DummyClassifier(strategy="most_frequent")
# Bundle preprocessing and modeling code in a pipeline
my_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                              ('model', clf)
                             ])

# Preprocessing of training data, fit model 
my_pipeline.fit(X_train, y_train)

# Preprocessing of test data, get predictions
y_predicted = my_pipeline.predict(X_test)
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_predicted)

0.7822222222222223

# Decision Tree Classifier

In [27]:
import numpy as np
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV
param_grid={
    'max_depth':[2,4,5,10,15,20],
    'criterion':['gini','entropy'],
    'max_leaf_nodes':[5,10,20,50,100],
    'min_samples_split':[5,10,15,20]
}

grid_tree = GridSearchCV(DecisionTreeClassifier(),param_grid,cv=5,scoring='accuracy',n_jobs=-1)



# Bundle preprocessing and modeling code in a pipeline
my_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                              ('grid_search', grid_tree)
                             ])

# Preprocessing of training data, fit model 
my_pipeline.fit(X_train, y_train)

print(grid_tree.best_estimator_)
print(np.abs(grid_tree.best_score_))

DecisionTreeClassifier(max_depth=4, max_leaf_nodes=20, min_samples_split=20)
0.8200000000000001


# K Nearest Neighbour

In [28]:
import numpy as np
from sklearn.model_selection import GridSearchCV
param_grid={
    'n_neighbors':[1,3,5,10],
    'algorithm':['auto', 'ball_tree', 'kd_tree', 'brute'],
    'weights':['uniform', 'distance'],
    'leaf_size':[2,10,20]
}

from sklearn.neighbors import KNeighborsClassifier
grid = GridSearchCV(KNeighborsClassifier(),param_grid,cv=3,scoring='accuracy',verbose=1,n_jobs=-1)

# Bundle preprocessing and modeling code in a pipeline
my_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                              ('grid_search', grid)
                             ])

# Preprocessing of training data, fit model 
my_pipeline.fit(X_train, y_train)

print(grid.best_estimator_)
print(np.abs(grid.best_score_))

Fitting 3 folds for each of 96 candidates, totalling 288 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  34 tasks      | elapsed:  1.1min
[Parallel(n_jobs=-1)]: Done 184 tasks      | elapsed:  5.8min
[Parallel(n_jobs=-1)]: Done 288 out of 288 | elapsed:  7.4min finished


Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  ('standardize',
                                                                   StandardScaler())]),
                                                  Index(['LIMIT_BAL', 'AGE', 'PAY_1', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5',
       'PAY_6', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4',
       'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3...
                                                                   OneHotEncoder(handle_unknown='ignore'))]),
                                                  Index(['SEX', 'EDUCATION', 'MARRIAGE'], dtype='object'))])),
                ('grid_search',
                 GridSearchCV(cv=3, esti

# Support Vector Machine

In [35]:
import numpy as np
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
param_grid={
    'C':[0.1,1,10,100],
    'gamma':[0.1,0.01,0.001,0.0001],
    'kernel':['rbf']
}

grid = GridSearchCV(SVC(),param_grid,cv=3,scoring='accuracy',verbose=1,n_jobs=-1)

# Bundle preprocessing and modeling code in a pipeline
my_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                              ('grid_search', grid)
                             ])

# Preprocessing of training data, fit model 
my_pipeline.fit(X_train, y_train)

print(grid.best_estimator_)
print(np.abs(grid.best_score_))

Fitting 3 folds for each of 16 candidates, totalling 48 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.


KeyboardInterrupt: 

In [36]:
from sklearn.svm import LinearSVC
from sklearn.model_selection import GridSearchCV
param_grid={
    'C':[0.1,1,10],
    'tol':[0.01,0.001,0.0001]
}

grid = GridSearchCV(LinearSVC(),param_grid,cv=3,scoring='accuracy',verbose=1,n_jobs=-1)

# Bundle preprocessing and modeling code in a pipeline
my_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                              ('grid_search', grid)
                             ])

# Preprocessing of training data, fit model 
my_pipeline.fit(X_train, y_train)

print(grid.best_estimator_)
print(np.abs(grid.best_score_))

Fitting 3 folds for each of 9 candidates, totalling 27 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  27 out of  27 | elapsed:   20.5s finished


LinearSVC(C=10, tol=0.01)
0.8066666666666666


C:\Users\jiahu_000\anaconda3\lib\site-packages\sklearn\svm\_base.py:975: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


# logistic Regression

In [60]:
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
param_grid={
    'solver':['lbfgs', 'liblinear', 'sag', 'saga'],
    'C':[0.01,0.1,1,10,100],
    'tol':[0.000001,0.0001,0.01,1,100]
}

grid_log = GridSearchCV(LogisticRegression(),param_grid,cv=3,scoring='accuracy',verbose=1, n_jobs=-1)

# Bundle preprocessing and modeling code in a pipeline
my_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                              ('grid_search', grid_log)
                             ])


# Preprocessing of training data, fit model 
my_pipeline.fit(X_train, y_train)

# displace best hyperparameters and best accuracy
print(grid_log.best_estimator_)
print(np.abs(grid_log.best_score_))

Fitting 3 folds for each of 100 candidates, totalling 300 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done  42 tasks      | elapsed:    9.5s
[Parallel(n_jobs=-1)]: Done 192 tasks      | elapsed:   33.6s
[Parallel(n_jobs=-1)]: Done 300 out of 300 | elapsed:   50.9s finished


LogisticRegression(C=0.1, solver='liblinear', tol=1e-06)
0.810625


# Random Forest

In [62]:
from sklearn.ensemble import RandomForestClassifier
import numpy as np
from sklearn.model_selection import GridSearchCV
param_grid={
    'criterion':['gini', 'entropy'],
    'max_depth':[1,10,100,1000]
}

grid_rf = GridSearchCV(RandomForestClassifier(),param_grid,cv=3,scoring='accuracy',verbose=1, n_jobs=-1)

# Bundle preprocessing and modeling code in a pipeline
my_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                              ('grid_search', grid_rf)
                             ])


# Preprocessing of training data, fit model 
my_pipeline.fit(X_train, y_train)

# displace best hyperparameters and best accuracy
print(grid_rf.best_estimator_)
print(np.abs(grid_rf.best_score_))

Fitting 3 folds for each of 8 candidates, totalling 24 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done  24 out of  24 | elapsed:   37.3s finished


RandomForestClassifier(criterion='entropy', max_depth=10)
0.81875


# XGBoost

In [41]:
# XGBoost
from xgboost import XGBClassifier
import numpy as np
from sklearn.model_selection import GridSearchCV
param_grid={
    'booster':['gbtree', 'gblinear', 'dart'],
    'max_depth':[1,10,100,1000]
}

grid = GridSearchCV(XGBClassifier(),param_grid,cv=3,scoring='accuracy',verbose=1, n_jobs=-1)

# Bundle preprocessing and modeling code in a pipeline
my_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                              ('grid_search', grid)
                             ])


# Preprocessing of training data, fit model 
my_pipeline.fit(X_train, y_train)

# displace best hyperparameters and best accuracy
print(grid.best_estimator_)
print(np.abs(grid.best_score_))

Fitting 3 folds for each of 12 candidates, totalling 36 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  36 out of  36 | elapsed:  1.4min finished


KeyError: 'base_score'

In [44]:
# XGBoost
from xgboost import XGBClassifier
import numpy as np
from sklearn.model_selection import GridSearchCV

clf = XGBClassifier()

# Bundle preprocessing and modeling code in a pipeline
my_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                              ('model', clf)
                             ])


# Preprocessing of training data, fit model 
my_pipeline.fit(X_train, y_train)

# Preprocessing of test data, get predictions
y_predicted = my_pipeline.predict(X_test)
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_predicted)


0.8138888888888889

# Naive Bayes

In [46]:
from sklearn.naive_bayes import GaussianNB
clf = GaussianNB()
import numpy as np
from sklearn.model_selection import GridSearchCV

# Bundle preprocessing and modeling code in a pipeline
my_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                              ('model', clf)
                             ])


# Preprocessing of training data, fit model 
my_pipeline.fit(X_train, y_train)

# Preprocessing of test data, get predictions
y_predicted = my_pipeline.predict(X_test)
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_predicted)

0.49822222222222223

# Ridge Classifier

In [47]:
from sklearn.linear_model import RidgeClassifier
clf = RidgeClassifier()
import numpy as np
from sklearn.model_selection import GridSearchCV

# Bundle preprocessing and modeling code in a pipeline
my_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                              ('model', clf)
                             ])


# Preprocessing of training data, fit model 
my_pipeline.fit(X_train, y_train)

# Preprocessing of test data, get predictions
y_predicted = my_pipeline.predict(X_test)
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_predicted)

0.7992222222222222

# Stochastic Gradient Descent Classifier

In [49]:
from sklearn.linear_model import SGDClassifier
clf = SGDClassifier()
import numpy as np
from sklearn.model_selection import GridSearchCV

# Bundle preprocessing and modeling code in a pipeline
my_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                              ('model', clf)
                             ])


# Preprocessing of training data, fit model 
my_pipeline.fit(X_train, y_train)

# Preprocessing of test data, get predictions
y_predicted = my_pipeline.predict(X_test)
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_predicted)

0.8048888888888889

# Neural Network

In [50]:
from sklearn.neural_network import MLPClassifier
clf = MLPClassifier()
import numpy as np
from sklearn.model_selection import GridSearchCV

# Bundle preprocessing and modeling code in a pipeline
my_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                              ('model', clf)
                             ])


# Preprocessing of training data, fit model 
my_pipeline.fit(X_train, y_train)

# Preprocessing of test data, get predictions
y_predicted = my_pipeline.predict(X_test)
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_predicted)

C:\Users\jiahu_000\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:585: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


0.8072222222222222